lib/diagnostics_channel.js

```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+const {
+  ObjectDefineProperty,
+  ObjectSetPrototypeOf,
+  ObjectValues,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      this._subscribers.push(subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        this._subscribers.push(subscription.onMessage.bind(subscription));
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    ObjectDefineProperty(this, 'name', {
+      value: name
```

Comment:
why use defineProperty here?

Reply:
I switched to just using plain properties. I went for `defineProperty` before just to prevent tampering with the value. Working in APM so long, I tend to code very defensively sometimes. 😅